src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractVersionControlService.java

```diff
@@ -67,4 +96,50 @@ public void addWebHookForParticipation(ProgrammingExerciseParticipation particip
             addWebHook(participation.getVcsRepositoryUrl(), ARTEMIS_SERVER_URL + PROGRAMMING_SUBMISSION_RESOURCE_API_PATH + participation.getId(), "Artemis WebHook");
         }
     }
+
+    @Override
+    public VcsRepositoryUrl copyRepository(String sourceProjectKey, String sourceRepositoryName, String targetProjectKey, String targetRepositoryName)
+            throws VersionControlException {
+        sourceRepositoryName = sourceRepositoryName.toLowerCase();
+        targetRepositoryName = targetRepositoryName.toLowerCase();
+        final String targetRepoSlug = targetProjectKey.toLowerCase() + "-" + targetRepositoryName;
+        // get the remote url of the source repo
+        final var sourceRepoUrl = getCloneRepositoryUrl(sourceProjectKey, sourceRepositoryName);
+        // get the remote url of the target repo
+        final var targetRepoUrl = getCloneRepositoryUrl(targetProjectKey, targetRepoSlug);
+        Repository targetRepo = null;
+        try {
+            // create the new target repo
+            createRepository(targetProjectKey, targetRepoSlug, null);
+            // clone the source repo to the target directory
+            targetRepo = gitService.getOrCheckoutRepositoryIntoTargetDirectory(sourceRepoUrl, targetRepoUrl, true);
+            // copy by pushing the source's content to the target's repo
+            gitService.pushSourceToTargetRepo(targetRepo, targetRepoUrl);
+        }
+        catch (InterruptedException | GitAPIException e) {
+            Path localPath = gitService.getDefaultLocalPathOfRepo(targetRepoUrl);
+            try {
+                if (targetRepo != null) {
+                    // delete the target repo if an error occurs
+                    gitService.deleteLocalRepository(targetRepo);
+                }
+                else {
+                    // or delete the folder if it exists
+                    FileUtils.deleteDirectory(localPath.toFile());
+                }
+            }
+            catch (IOException ex) {
+                // ignore
+                log.error("Could not delete directory of the failed cloned repository in: " + localPath);
+            }
+            throw new BitbucketException("Could not copy repository " + sourceRepositoryName + " to the target repository " + targetRepositoryName, e);
```

Comment:
Should this rather be a `AbstractVersionControlServiceException`?

Reply:
of course! Thanks, I've fixed that.